Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs

```diff
@@ -71,30 +71,29 @@ INotificationPublisher notificationPublisher
       );
     }
 
-    public async Task<DocumentAfterParsing> LoadAsync(DocumentTextBuffer textDocument,
+    public async Task<DocumentAfterResolution> LoadAsync(DocumentAfterParsing documentAfterParsing,
       CancellationToken cancellationToken) {
 #pragma warning disable CS1998
       return await await Task.Factory.StartNew(
-        async () => LoadInternal(textDocument, cancellationToken), cancellationToken,
+        async () => LoadInternal(documentAfterParsing, cancellationToken), cancellationToken,
 #pragma warning restore CS1998
         TaskCreationOptions.None, ResolverScheduler);
     }
 
-    private DocumentAfterParsing LoadInternal(DocumentTextBuffer textDocument,
+    private DocumentAfterResolution LoadInternal(DocumentAfterParsing documentAfterParsing,
       CancellationToken cancellationToken) {
-      var errorReporter = new DiagnosticErrorReporter(textDocument.Text, textDocument.Uri);
-      statusPublisher.SendStatusNotification(textDocument, CompilationStatus.ResolutionStarted);
-      var program = parser.Parse(textDocument, errorReporter, cancellationToken);
-      var documentAfterParsing = new DocumentAfterParsing(textDocument, program, errorReporter.GetDiagnostics(textDocument.Uri));
-      if (errorReporter.HasErrors) {
-        statusPublisher.SendStatusNotification(textDocument, CompilationStatus.ParsingFailed);
-        return documentAfterParsing;
-      }
 
-      var compilationUnit = symbolResolver.ResolveSymbols(textDocument, program, out _, cancellationToken);
-      var symbolTable = symbolTableFactory.CreateFrom(program, compilationUnit, cancellationToken);
+      var textDocument = documentAfterParsing.TextDocumentItem;
+      var parsedProgram = documentAfterParsing.Program;
+      var clonedProgram = new Program(parsedProgram.Name,
```

Comment:
Oops the cloning removes all the ranges and owned tokens, that means that the parsed document will be useless for formatting... I don't think we can get around solving https://github.com/dafny-lang/dafny/issues/3290 now

Reply:
Can we not just clone those as well? Not cloning them seems like a bug

Reply:
Yes, but since they are not part of the constructor, we might forget about them, either in current cloners, in future cloners, or in future language constructs. Anyway, we can add the range manually for now.